src/pymor/algorithms/rand_la.py

```diff
@@ -2,21 +2,266 @@
+
+        if len(self._testvecs) < num_testvecs:
+            self._draw_test_vector(num_testvecs - len(self._testvecs))
+
+        W, Q = self._testvecs[:num_testvecs].copy(), self._find_range(basis_size)
+        W -= Q.lincomb(Q.inner(W, self.range_product).T)
+        return c * np.max(W.norm(self.range_product))
+
+    def estimate_error(self, basis_size, num_testvecs=20, p_fail=1e-14):
+        r"""Randomized a posteriori error estimator for a given basis size.
+
+        This implements the a posteriori error estimator from :cite:`BS18` (Definition 3.1).
+
+        The error is given by
+
+        .. math::
+            \epsilon_{\mathrm{est}}=c_{\mathrm{est}}\cdot\max_{\omega\in\Omega}
+            \lVert (I-QQ^T)A\omega\rVert_{T\rightarrow S}
+
+        with
+
+        .. math::
+            c_{\mathrm{est}}
+            =\frac{1}{\sqrt{2\lambda_{\mathrm{min}}}\operatorname{erf}^{-1}
+            \left(\left(\frac{\texttt{p_fail}}{n}\right)^{1/\texttt{num_testvecs}}\right)},
+
+        where :math:`\Omega` is a set of `num_testvecs` random vectors, :math:`S` denotes the inner
+        product of the range, :math:`T` denotes inner product of the source of :math:`A` and
+        :math:`n=\min\{\dim(S),\,\dim(T)\}`.
+
+        Parameters
+        ----------
+        basis_size
+            The size of the basis.
+        num_testvecs
+            Number of test vectors for error estimation.
+        p_fail
+            Maximum failure probabilty of the error estimate.
+
+        Returns
+        -------
+        err
+            The approximate error of the basis.
+        """
+        assert isinstance(basis_size, int) and basis_size > 0
+        if basis_size > min(self.A.source.dim, self.A.range.dim):
+            self.logger.warning('Requested basis is larger than the rank of the operator!')
+            self.logger.info('Proceeding with maximum operator rank.')
+            basis_size = min(self.A.source.dim, self.A.range.dim)
+        assert 0 < num_testvecs and isinstance(num_testvecs, int)
+        assert 0 < p_fail < 1
+
+        err = self._estimate_error(basis_size, num_testvecs, p_fail)
+        self.logger.info(f'Estimated error (basis dimension {basis_size}): {err:.5e}.')
+        return err
+
+    def _extend_basis(self, k):
+        W = self.A.source.random(k, distribution='normal')
+        if self.complex:
+            W += 1j * self.A.source.random(k, distribution='normal')
+
+        offset = len(self._Q[0])
+        self._Q[0].append(self.A.apply(W))
+        gram_schmidt(self._Q[0], self.range_product, offset=offset, copy=False)
+
+        for i in range(self.subspace_iterations):
+            i = 2*i + 1
+
+            k = len(self._Q[i-1]) - offset  # check if GS removed vectors
+            offset = len(self._Q[i])
+            self._Q[i].append(self.source_product.apply_inverse(
+                (self._adjoint_op.apply(self.range_product.apply(self._Q[i-1][-k:])))))
```

Comment:
I think this is wrong: let $S$ and $R$ be the matrices for the source and range product and $A$ the matrix of the operator. Then the adjoint of $A$ *with respect to inner products on source and range* is given by $S^{-1}A^TR$:
$$(S^{-1}A^TR w, v)_S = (SS^{-1}A^TR w, v) = (A^TRw, v) = (Rw, Av) = (w, Av)_R.$$
What we need in the GHEP case is to specify that the passed operator is self-adjoint w.r.t. the $E$-inner product. Thus, we would pass $E$ both as `source_product` and `range_product` and only have
```python
            self._Q[i].append(self._adjoint_op.apply(self._Q[i-1][-k:])
```
To avoid having to check whether an adjoint operator was given here, you could write above
```python
AdjointOperator(A, range_product=range_product, source_product=source_product)
```

Reply:
> I think this is wrong: let S and R be the matrices for the source and range product and A the matrix of the operator. Then the adjoint of A _with respect to inner products on source and range_ is given by S−1ATR: (S−1ATRw,v)S=(SS−1ATRw,v)=(ATRw,v)=(Rw,Av)=(w,Av)R.

I'm not sure I follow. Don't lines 164 and 165 exactly amount to $$Sˆ{-1}AˆTRx$$?


What we need in the GHEP case is to specify that the passed operator is self-adjoint w.r.t. the E-inner product. Thus, we would pass E both as `source_product` and `range_product` and only have
> 
> ```python
>             self._Q[i].append(self._adjoint_op.apply(self._Q[i-1][-k:])
> ```

In the GHEP case RRF is called without any source or range products now. In that case they are set as `IdentityOperators`, so lines 164 and 165 are equivalent to your suggestion.

Reply:
Wow! At last, this makes the basis extension readable 🎉

Reply:
I'm not sure what to do now. With that syntax, we actually don't even need that `self_adjoint` kwarg anymore right? The amount of overhead caused by `AdjointOperator` is neglible, right?

Reply:
> > I think this is wrong: let S and R be the matrices for the source and range product and A the matrix of the operator. Then the adjoint of A _with respect to inner products on source and range_ is given by S−1ATR: (S−1ATRw,v)S=(SS−1ATRw,v)=(ATRw,v)=(Rw,Av)=(w,Av)R.
> 
> I'm not sure I follow. Don't lines 164 and 165 exactly amount to S−1ATRx?

Yes.

> 
> What we need in the GHEP case is to specify that the passed operator is self-adjoint w.r.t. the E-inner product. Thus, we would pass E both as `source_product` and `range_product` and only have
> 
> > ```python
> >             self._Q[i].append(self._adjoint_op.apply(self._Q[i-1][-k:])
> > ```
> 
> In the GHEP case RRF is called without any source or range products now. In that case they are set as `IdentityOperators`, so lines 164 and 165 are equivalent to your suggestion.

Yes, but orthonormalization happens w.r.t. the Euclidean inner product. As I said, not sure how much this matters for intermediate steps, however, we surely want the final basis to be `E`-orthogonal.

Reply:
> I'm not sure what to do now. With that syntax, we actually don't even need that `self_adjoint` kwarg anymore? 

We do: If we were to pass $E^{-1}A$ as operator and $E$ as source and range product, we would IMHO get the mathematically right result. However, we would always use $E^{-1}AE^{-1}E$ as the adjoint operator, which would amount to solving an additional linear system. This can be avoided telling the range finder that the operator is self-adjoint.